README.md
# skyrts

## redis 实时统计系统

- 按秒，分钟，小时，天，周，月，年统计
- 按每（分钟、小时等）每小时，天，周，月，年聚合
- 自由设置时间盒子
- 基础聚合值 Min, Max, Avg, Sum, Range（Max-Min）,Count ,First(首次进入值) ,Last(最后进入值)
- 高级聚合值 M2(二阶中心矩), Var.P（M2/Count）,Var.S（M2/(Count-1)), STD.P(Sqrt(Var.P)), STD.S(Sqrt(Var.S))

```javascript
const rts = require('skyrts')
const Pack = require('./package.json')

rts({
  redis: redis, // 异步redis引用
  redisAsync: redis, // 同步redis引用
  gran: '1s, 5m, 1h, 1d, 1w, 1M, 1y', // 维度的梯度
  points: 1000, // 多少点循环记录，默认500个点
  prefix: Pack.name // 需要一个前缀，区分多项目
})
```

- 如果需要 按季度记录，gran 中增加 3M
- 如果要经常统计全局总值，gran 中增加 9999y 9999 年

### 主要方法

```javaScript
// 埋点
rts.record(name, num = 1, statistics = ['sum', 'avg'], aggregations, timestamp, callback)
```

- 异步记录函数，一般常用此方法
- name 准备存放的字符串，最终形式是 '_rts_'+options.prefix + name
- num 默认 1 sum+=1 否则 sum+=num
- statistics 默认统计 count 计数，sum 总和，avg 平均值
- 启用 avg 后，同时计算，m2 二阶中心矩，first 首次进入值，last 最后进入值
- aggregations {Array} dy (day in week each year), hm(hour of day each month) null 就不处理聚合
- timestamp 可以补历史记录，默认是当前服务器时间长整型
- callback 一般不用

```javaScript
// 去重埋点 标准误差为0.81%
rts.recordUnique(name, uniqueId, statistics, aggregations, timestamp, callback)
```

- 异步唯一值记录函数，一般常用此方法，记录后不会立刻生效，会在 options.interval || 60000 60 秒刷新唯一值
- name 准备存放的字符串，最终形式是 '_rts_'+options.prefix + name
- uniqueId 唯一标识字符串，可以为数组形式，多个唯一值
- statistics 默认为空 记录唯一值，如果是 ['sum', 'avg']形式 调用普通 record 函数
- aggregations {Array} dy (day in week each year), hm(hour of day each month) null 就不处理聚合
- timestamp 可以补历史记录，默认是当前服务器时间长整型

```javaScript
// 异步获取埋点时间序列
rts.getStat(type, name, granCode, fromDate, toDate, callback)
```

- 不常用，看下面同步方法

```javaScript
// 同步获取埋点时间序列
await rts.getStatAsync(type, name, granCode, fromDate, toDate)
```

- type sum, min, max, avg, count, uni 唯一值情况
- name key 的名称
- granCode 开始 rts 设置的 gran: '1s, 5m, 1h, 1d, 1w, 1M, 1y'
- fromDate 开始时间 Date 类型
- toDate 结束时间 Date 类型
- 可以使用 meeko 中的 data.offset()

```javaScript
const offset = function (interval, number) {
  /**
   * @function [dateAdd&offset]
   * @description 日期偏移操作.
   * @memberof Date_prototype#
   * @param {string} interval - 年月日时分秒周季 yMdhnswq.
   * @param {int} number - 时间间隔 可正负.
   * @return {string} 返回 得到日期年月日等加数字后的日期.
   * @example
   * $.now().offset('y',-1)
   * // 2018-6-1 10:19:01
   */
}
```

### 实战例子

![demo1](https://github.com/kongnet/skybase/raw/master/screenShot/demo1.png)


CHANGELOG.md
## [0.1.10](https://github.com/kongnet/skyrts/compare/v0.1.9...v0.1.10) (2022-03-17)




## [0.1.9](https://github.com/kongnet/skyrts/compare/v0.1.8...v0.1.9) (2022-03-14)




## [0.1.8](https://github.com/kongnet/skyrts/compare/v0.1.7...v0.1.8) (2022-03-11)




## [0.1.7](https://github.com/kongnet/skyrts/compare/v0.1.6...v0.1.7) (2022-03-11)




## [0.1.6](https://github.com/kongnet/skyrts/compare/5608f67873f197f416e28b27b87001a7cf16c6de...v0.1.6) (2022-03-11)


### :art:

* init ([5608f67](https://github.com/kongnet/skyrts/commit/5608f67873f197f416e28b27b87001a7cf16c6de))

### :bug:

* 修复只有count计数时候的问题 ([f8d0572](https://github.com/kongnet/skyrts/commit/f8d057249bd22e745345d017d96f184a97617953))

### :memo:

* 增强readme ([4efa446](https://github.com/kongnet/skyrts/commit/4efa4468870b6c1806b5258a65455896b73666b3))

### :sparkles:

* 增加加载额外hash函数loadOneScript ([1d03eb2](https://github.com/kongnet/skyrts/commit/1d03eb2e5c1573da69c8c5bc30b50114f47432f2))
* avg.lua增加first last 和 m2二阶中间矩 ([77bed00](https://github.com/kongnet/skyrts/commit/77bed007deebc90a53549f868e29ef0cac5a91b3))





util/util.js
// vim: sw=4
exports.normalPeriod = {
  s: 1000,
  m: 60 * 1000,
  h: 60 * 60 * 1000,
  d: 24 * 60 * 60 * 1000,
  w: 7 * 24 * 60 * 60 * 1000,
  M: 30 * 24 * 60 * 60 * 1000,
  y: 365 * 24 * 60 * 60 * 1000
}

exports.getUnitDesc = function (str) {
  str = str.trim()
  let match = str.match(/(\d+)([smhdwMy])/)
  if (!match) throw new Error('Bad period descriptor:' + str)
  let num = parseInt(match[1]) || 1
  let unit = match[2]
  let period = num * exports.normalPeriod[unit]
  return [period, str, num, unit]
}

function getFixedTime (timestamp) {
  let d = new Date()
  // return timestamp;
  return timestamp - d.getTimezoneOffset() * 60 * 1000
}

function restoreFixedTime (timestamp) {
  let d = new Date()
  // return timestamp;
  return timestamp + d.getTimezoneOffset() * 60 * 1000
}

// the time point on the report
exports.getKeyTime = function (unitDesc, timestamp) {
  if (!timestamp) timestamp = Date.now()
  let num = unitDesc[2]
  let unit = unitDesc[3]
  let period = unitDesc[0]
  let v
  if (unit === 'y') {
    let year = (new Date(timestamp)).getFullYear()
    year = year - year % num
    v = (new Date(year, 0, 1)).getTime()
  } else if (unit === 'M') {
    let d = new Date(timestamp)
    year = d.getFullYear()
    let month = d.getMonth()
    month = month - month % num
    v = (new Date(year, month, 1)).getTime()
    // TODO week
  } else {
    timestamp = getFixedTime(timestamp)
    v = timestamp - (timestamp % period)
    v = restoreFixedTime(v)
  }
  return v
}

exports.getExpireTime = function (unitDesc, timestamp) {
  if (!timestamp) timestamp = Date.now()
  let num = unitDesc[2]
  let unit = unitDesc[3]
  let period = unitDesc[0]
  let v
  if (unit === 'y') {
    let year = (new Date(timestamp)).getFullYear()
    year = year - year % num
    v = (new Date(year + 1, 0, 1)).getTime()
  } else if (unit === 'M') {
    let d = new Date(timestamp)
    year = d.getFullYear()
    let month = d.getMonth()
    month = month - month % num
    v = (new Date(year, month + 1, 1)).getTime()
    // TODO week
  } else {
    timestamp = getFixedTime(timestamp)
    v = timestamp - (timestamp % period) + period
    v = restoreFixedTime(v)
  }
  return v
}

exports.getGranId = function (unitDesc, timestamp) {
  if (!timestamp) timestamp = Date.now()
  let num = unitDesc[2]
  let unit = unitDesc[3]
  let period = unitDesc[0]
  let v
  let d = new Date(timestamp)
  if (unit === 'y') {
    v = d.getFullYear() / num
  } else if (unit === 'M') {
    v = (d.getFullYear() * 100 + d.getMonth() + 1) / num
    // TODO week
  } else {
    timestamp = getFixedTime(timestamp)
    v = timestamp / period
  }
  return Math.floor(v)
}

exports.getAggrId = function (desc, timestamp) {
  let d = timestamp ? new Date(timestamp) : new Date()
  if (desc[0] === 'h') {
    return d.getHours()
  } else if (desc[0] === 'd') {
    return d.getDay()
  }
  throw new Error('unknow aggr desc ' + desc)
}

exports.getAggrGroupId = function (desc, timestamp) {
  let v = 'all'
  let d = new Date(timestamp)
  if (desc[1] === 'm') {
    v = d.getFullYear() * 100 + d.getMonth()
  } else if (desc[1] === 'q') {
    v = d.getFullYear() * 10 + Math.ceil(d.getMonth() / 3)
  } else if (desc[1] === 'y') {
    v = d.getFullYear()
  }
  return desc + '.' + v
}

exports.getTimePeriod = function (gran, points) {
  if (typeof gran === 'string') {
    gran = exports.getUnitDesc(gran)
  }
  return points * gran[0]
}


index.js
let CHANNEL = '_rts_.record'
let util = require('./util/util')
let fs = require('fs')
let scripts = {}
let scriptSha1s = {}
let redis = null
let redisSync = null
let defaultFileArr = ['avg', 'max', 'min', 'update_pf']
defaultFileArr.forEach(function (key) {
  scripts[key] = fs.readFileSync(__dirname + '/lua/' + key + '.lua', 'utf-8')
})

function loadScripts (client) {
  //兼容老的模式，在出错的时候再次加载
  for (let scriptName in scripts) {
    ;(function (scriptName) {
      client.script('load', scripts[scriptName], function (err, sha1) {
        if (sha1) scriptSha1s[scriptName] = sha1
      })
    })(scriptName)
  }
}

async function loadDefaultScripts () {
  // 兼容老的
  const multi = redisSync.multi()
  for (let key in scripts) {
    await multi.script('load', scripts[key])
  }
  let results = await multi.exec()
  results.forEach(
    (x, idx) => (scriptSha1s[defaultFileArr[idx]] = results[idx][1])
  )
  //console.log(scriptSha1s)
}

async function loadOneScript (key, filePath) {
  try {
    if (!key) {
      console.log('loadScriptErr:', '必须有脚本的对应的ey')
      return -1
    }
    scripts[key] = fs.readFileSync(filePath, 'utf-8')
    let multi = await redisSync.multi()
    await multi.script('load', scripts[key])
    let results = await multi.exec()
    if (results[0][1]) scriptSha1s[key] = results[0][1]
    return results[0][1]
  } catch (e) {
    console.log('loadScriptErr:', e)
    return -1
  }
}
function evalScript (client, scriptName, keys, args, callback) {
  // console.log(`---------client.script('EXISTS', scriptName)------------${client.script('EXISTS', scriptName)}`)
  if (scriptSha1s[scriptName]) {
    args = [scriptSha1s[scriptName], keys.length]
      .concat(keys || [])
      .concat(args || [])
    if (callback) {
      args.push(callback)
    }
    client.evalsha.apply(client, args)
  } else {
    args = [scripts[scriptName], keys.length]
      .concat(keys || [])
      .concat(args || [])
    if (callback) {
      args.push(callback)
    }
    client.eval.apply(client, args)
  }
}

let aggrvals = {
  h: [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23
  ],
  d: [0, 1, 2, 3, 4, 5, 6]
}

function log (err, results) {
  if (err) console.log(err.stack || err)
}

/**
 * Create a rts instance
 *
 * @param {Object} options
 *          redis:
 *          gran: comma seprate string
 *          points: how many points store
 *          interval: update interval for pflog
 *          prefix: a prefix for all keys
 *
 */
exports = module.exports = async function rts (options) {
  redis = options.redis
  redisSync = options.redisSync

  let granularities = options.gran || '5m, 1h, 1d, 1w'

  let prefix = options.prefix || ''

  let points = options.points || 500

  let interval = options.interval || 60000

  //loadScripts(redis)
  await loadDefaultScripts(redisSync)
  prefix = '_rts_' + prefix

  let keyPFKeys = prefix + ':pfkeys'

  granularities = granularities.split(',').map(util.getUnitDesc)
  let granMap = {}
  granularities.forEach(function (granInfo) {
    granMap[granInfo[1]] = granInfo
  })

  function getGranKey (name, gran, timestamp) {
    let granId = util.getGranId(gran, timestamp)
    return [prefix, name, gran[1], granId].join(':')
  }

  function getAggrGruopKey (name, aggr, timestamp) {
    let gid = util.getAggrGroupId(aggr, timestamp)
    return [prefix, name, 'aggr', gid].join(':')
  }

  function getAggrKey (name, aggr, timestamp) {
    let aggrId = util.getAggrId(aggr, timestamp)
    return getAggrGruopKey(name, aggr, timestamp) + '.' + aggrId
  }

  // some thing's some statics value in some granularity at some time
  function setValue (name, stat, value, gran, timestamp, callback) {
    if (typeof gran === 'string') {
      gran = util.getUnitDesc(gran)
    }
    let key = getGranKey(name, gran, timestamp)
    redis.hset(key, stat, value, callback)
  }

  /**
   * record behavior.
   * @param statistics {Array} sum, max, min, count, avg
   * @param aggregations {Array} dy (day in week each year), hm(hour of day each month).
   *
   */
  function record (
    name,
    num = 1,
    statistics = ['sum', 'avg'],
    aggregations,
    timestamp,
    callback
  ) {
    if (statistics.includes('count')) {
      if (statistics.includes('avg')) {
        statistics = statistics.filter(x => x !== 'count')
      } else {
        statistics.push('avg')
      }
    }
    timestamp = +new Date(timestamp) || Date.now()
    function recordStats (key) {
      statistics &&
        statistics.forEach(function (stat) {
          if (stat === 'sum') {
            multi.hincrbyfloat(key, 'sum', num) // NOTICE: 必须是整数
          } else if (scripts[stat]) {
            evalScript(multi, stat, [key, num], [], function (err, result) {
              try {
                if (err) throw err
                if (result && result.indexOf('ERR') === 0) {
                  // console.log(scripts[stat]);
                  throw new Error(result)
                }
              } catch (e) {
                console.log(`-----------evalScriptERR------------ ${e}`)
                // throw new Error(e)
                // 万一出现问题 从新load一次lua文件
                loadScripts(redis)
              }
            })
          }
        })
    }

    let multi = redis.multi()
    granularities.forEach(function (gran) {
      let key = getGranKey(name, gran, timestamp)
      recordStats(key)
      let unitPeriod = gran[0]
      multi.expire(key, (points * unitPeriod) / 1000)
    })

    if (aggregations) {
      for (let i = 0; i < aggregations.length; i++) {
        let aggr = aggregations[i]
        let key = getAggrKey(name, aggr, timestamp)
        recordStats(key)
      }
    }

    multi.exec(callback || log)
  }

  /**
   * record unique access, like unique user of a period time.
   * @param {string | Array} uniqueId one or some uniqueId to be stats
   */
  function recordUnique (
    name,
    uniqueId,
    statistics,
    aggregations,
    timestamp,
    callback
  ) {
    timestamp = timestamp || Date.now()
    // normal record
    if (statistics) {
      let num = Array.isArray(uniqueId) ? uniqueId.length : 1
      record(name, num, statistics, aggregations, timestamp)
    }
    // record unique
    let multi = redis.multi()
    granularities.forEach(function (gran) {
      let key = getGranKey(name, gran, timestamp)
      let expireTime = util.getExpireTime(gran, timestamp)
      let pfkey = key + ':pf'
      multi.hset(keyPFKeys, pfkey, expireTime)
      if (Array.isArray(uniqueId)) {
        multi.pfadd.apply(multi, [pfkey].concat(uniqueId))
      } else {
        multi.pfadd(pfkey, uniqueId)
      }
      // recordStats(key);
      let unitPeriod = gran[0]
      multi.hincrby(key, 'uni', 0)
      multi.expire(key, (points * unitPeriod) / 1000)
    })
    multi.exec(callback || log)
  }
  /**
   * get results of the stats
   *
   * @param {String} type sum, min, max, avg, count, uni
   */
  async function getStatAsync (
    type,
    name,
    granCode,
    fromDate,
    toDate,
    callback
  ) {
    if (!granCode) throw new Error('granCode is required')
    if (!callback && typeof toDate === 'function') {
      callback = toDate
      toDate = Date.now()
    }
    let gran = granMap[granCode] || util.getUnitDesc(granCode)
    if (!gran) throw new Error('Granualrity is not defined ' + granCode)
    if (fromDate instanceof Date) fromDate = fromDate.getTime()
    if (toDate instanceof Date) toDate = toDate.getTime()

    toDate = toDate || Date.now()
    fromDate = fromDate || toDate - util.getTimePeriod(gran, points)
    let unitPeriod = gran[0]
    let multi = await redisSync.multi()
    if (!multi) throw new Error('redis multi error')
    let _points = []
    for (let d = fromDate; d <= toDate; d += unitPeriod) {
      let key = getGranKey(name, gran, d)
      _points.push(util.getKeyTime(gran, d))
      await multi.hget(key, type)
    }
    let results = await multi.exec()
    let merged = []
    for (let i = 0, l = _points.length, p; i < l; i++) {
      p = _points[i]
      if (Array.isArray(results[i])) {
        merged[i] = [p, Number(results[i][1])]
      } else {
        merged[i] = [p, Number(results[i])]
      }
    }
    return {
      step: unitPeriod,
      unitType: gran[3],
      data: merged
    }
  }
  /**
   * get results of the stats
   *
   * @param {String} type sum, min, max, avg, count, uni
   */
  function getStat (type, name, granCode, fromDate, toDate, callback) {
    if (!granCode) throw new Error('granCode is required')
    if (!callback && typeof toDate === 'function') {
      callback = toDate
      toDate = Date.now()
    }
    let gran = granMap[granCode] || util.getUnitDesc(granCode)
    if (!gran) throw new Error('Granualrity is not defined ' + granCode)
    if (fromDate instanceof Date) fromDate = fromDate.getTime()
    if (toDate instanceof Date) toDate = toDate.getTime()

    toDate = toDate || Date.now()
    fromDate = fromDate || toDate - util.getTimePeriod(gran, points)
    let unitPeriod = gran[0]
    let multi = redis.multi()
    let _points = []
    for (let d = fromDate; d <= toDate; d += unitPeriod) {
      let key = getGranKey(name, gran, d)
      _points.push(util.getKeyTime(gran, d))
      multi.hget(key, type)
    }
    multi.exec(function (err, results) {
      if (err) return callback(err)
      let merged = []
      for (let i = 0, l = _points.length, p; i < l; i++) {
        p = _points[i]
        if (Array.isArray(results[i])) {
          merged[i] = [p, Number(results[i][1])]
        } else {
          merged[i] = [p, Number(results[i])]
        }
      }
      callback(null, {
        step: unitPeriod,
        unitType: gran[3],
        data: merged
      })
    })
  }

  function aggrstat (type, name, aggr, date, callback) {
    if (!callback && typeof date === 'function') {
      callback = date
      date = Date.now()
    }
    let vals = aggrvals[aggr[0]]
    let multi = redis.multi()
    let gkey = getAggrGruopKey(name, aggr, date)
    for (let i = 0, l = vals.length; i < l; i++) {
      let key = gkey + '.' + vals[i]
      multi.hget(key, type)
    }
    multi.exec(function (err, results) {
      if (err) return callback(err)
      callback(
        null,
        results.map(function (result, i) {
          return [i, Number(result)]
        })
      )
    })
  }

  function updateHyperLogLog () {
    redis.hgetall(keyPFKeys, function (err, result) {
      if (err) return log(err)
      let now = Date.now()
      let multi = redis.multi()
      let expireTime, key
      for (let pfkey in result) {
        expireTime = Number(result[pfkey])
        key = pfkey.substring(0, pfkey.length - 3)
        evalScript(multi, 'update_pf', [pfkey, key], [])
        if (expireTime < now) {
          multi.hdel(keyPFKeys, pfkey)
          multi.del(pfkey)
        }
      }
      multi.exec(log)
    })
  }

  let updateTimer = setInterval(updateHyperLogLog, interval)

  function stop () {
    clearInterval(updateTimer)
  }

  return {
    record,
    recordUnique,
    stat: getStat, // deprecated
    getStat,
    getStatAsync,
    aggrstat,
    unique: getStat.bind(null, 'uni'),
    uni: getStat.bind(null, 'uni'),
    sum: getStat.bind(null, 'sum'),
    count: getStat.bind(null, 'count'),
    avg: getStat.bind(null, 'avg'),
    max: getStat.bind(null, 'max'),
    min: getStat.bind(null, 'min'),
    aggrsum: aggrstat.bind(null, 'sum'),
    aggrcount: aggrstat.bind(null, 'count'),
    aggravg: aggrstat.bind(null, 'avg'),
    aggrmax: aggrstat.bind(null, 'max'),
    aggrmin: aggrstat.bind(null, 'min'),
    setValue,
    getTimePeriod: util.getTimePeriod,
    stop,
    loadScripts,
    loadOneScript,
    scriptSha1s,
    scripts
  }
}

exports.getTimePeriod = util.getTimePeriod
